shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java

```diff
@@ -266,6 +268,38 @@ public void unspecifiedAction(int inputClientId) {
         Arrays.asList(inputClientId, "TextInputAction.unspecified"));
   }
 
+  public void performPrivateCommand(int inputClientId, String action, Bundle data) {
+    HashMap<Object, Object> json = new HashMap<>();
+    json.put("action", action);
+    if (data != null) {
+      HashMap<String, Object> dataMap = new HashMap<>();
+      Set<String> keySet = data.keySet();
+      for (String key : keySet) {
+        Object value = data.get(key);
+        if (value instanceof byte[]) {
```

Comment:
Didn't read the framework part yet. How would the framework figure out what type to decode into given the serialized form? Also some of this is already done via the JSONMethodCodec / JSONUtil classes. You may be able to reuse some parts.

Reply:
The command is a client app specific command between client app and the input method. They made the agreement what command and data format that the client app can send and the input method can handle. If the command isn't supported by the input method, nothing happens. If the data getting back from the input method is not the format that the client wants, the data is not valid. In Crowdsource Gboard project the workflow like this:
From Crowdsource (client app), it sends app private command to Gboard (input method) like this: 
{
  ‘action’: 'com.google.android.apps.village.boond.GET_INPUT_CONTEXT',
  ‘data’: ‘’
}

From Gboard, it performs the app private command to Crowdsource like this:
{
  ‘action’: 'com.google.android.apps.village.boond.GET_INPUT_CONTEXT',
  ‘data’: {
    ‘input_context’: xxxxxxxxxxxxx(byte array)
  }
}

Let me explore JSONMethodCodec / JSONUtil a little bit and adjust the implementation here.

Reply:
@xster  How to use JSONMethodCodec / JSONUtil in here? 
JSONMethodCodec is used for encoding and decoding ByteBuffer and it is not required in here. 
JSONUtil#wrap() can be used for wrapping into JSONObject, but in here we are trying to get the data type from Bundle data. In order to get the data with bundle#get<DataType>(), it has to know the data type.

Reply:
Did you mean that for all users of performPrivateCommand, the "schema" for the data is always expected to be tightly mapped to the action?

Reply:
Right. The "schema" of the data is mapped to the action.

Reply:
Understood. LGTM